crates/rattler_solve/src/resolvo/conda_util.rs

```diff
@@ -1,17 +1,344 @@
+struct MaxSolvable {
+    // The version of the dependency
+    // Only sort by version
+    version: Option<TrackedFeatureVersion>,
+    solvable_id: SolvableId,
+}
+
+impl MaxSolvable {
+    fn new(version: Option<TrackedFeatureVersion>, solvable_id: SolvableId) -> Self {
+        Self {
+            version,
+            solvable_id,
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+struct TrackedFeatureVersion {
+    version: Version,
+    tracked_features: bool,
+}
+
+impl TrackedFeatureVersion {
+    fn new(version: Version, tracked_features: bool) -> Self {
+        Self {
+            version,
+            tracked_features,
+        }
+    }
+}
+
+impl Ord for TrackedFeatureVersion {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.version.cmp(&other.version) {
+            Ordering::Equal => match (self.tracked_features, other.tracked_features) {
+                (true, false) => Ordering::Less,
+                (false, true) => Ordering::Greater,
+                _ => Ordering::Equal,
+            },
+            other => other,
+        }
+    }
+}
+
+impl PartialOrd for TrackedFeatureVersion {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A struct that calculates the score for each solvable based on the dependencies
+struct DependencyScores {
+    max_map: HashMap<NameId, Vec<MaxSolvable>>,
+}
+
+impl DependencyScores {
+    fn from_dependencies(
+        shared_dependencies: Vec<(SolvableId, HashMap<NameId, VersionSetId>)>,
+        solver: &SolverCache<CondaDependencyProvider<'_>>,
+        highest_version_cache: &mut HashMap<VersionSetId, Option<(Version, bool)>>,
+    ) -> Self {
+        // Map with the maximum version per name
+        let mut max_map = HashMap::new();
+        for (solvable, dependencies) in shared_dependencies {
+            for (name, version_set_id) in dependencies {
+                let version = find_highest_version(version_set_id, solver, highest_version_cache)
+                    .map(|v| TrackedFeatureVersion::new(v.0, v.1));
+                // Update the max version for the name
+                let max_solvable = MaxSolvable::new(version, solvable);
+                max_map
+                    .entry(name)
+                    .and_modify(|v: &mut Vec<MaxSolvable>| {
+                        v.push(max_solvable.clone());
+                    })
+                    .or_insert_with(|| vec![max_solvable]);
+            }
+        }
+
+        // Sort all vectors of dependencies by version
+        for max_solvables in max_map.values_mut() {
+            max_solvables.sort_by(|a, b| a.version.cmp(&b.version));
+            // dbg!(max_solvables);
+        }
+
+        Self { max_map }
+    }
+
+    /// Per dependency, score the solvables based on the highest version of the dependency
+    fn score_solvables(&self) -> HashMap<SolvableId, u32> {
+        let mut scores = HashMap::new();
```

Comment:
Maybe we can preallocate all the entries with `with_capacity`?

Reply:
Good one!